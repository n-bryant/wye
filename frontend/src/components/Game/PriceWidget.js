import React from "react";
import PropTypes from "prop-types";

import { withStyles } from "@material-ui/core/styles";
import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../../lib/classNamePrefix";

import Typography from "@material-ui/core/Typography";

import styles from "./PriceWidget.styles";

/**
 * renders a widget containing price details
 */
export class PriceWidget extends React.Component {
  constructor(props) {
    super(props);
    this.classnames = PriceWidget.classnames(props);
  }

  /**
   * renders the appropriate price block based on whether
   * a game is free or on sale
   */
  getPrice() {
    const { data, skinny } = this.props;
    if (data.freeToPlay) {
      return (
        <Typography
          className={this.classnames.element("priceFree")}
          variant="h2"
        >
          Free to Play
        </Typography>
      );
    } else if (data.onSale) {
      return (
        <div
          className={this.classnames.element("priceDetailsContainer", {
            withDiscount: true
          })}
        >
          <Typography
            className={this.classnames.element("discountPercent", {
              skinny
            })}
            variant="h1"
          >
            -{data.discountPercentage}%
          </Typography>
          <div className={this.classnames.element("priceDetails")}>
            <Typography
              className={this.classnames.element("initialPrice")}
              variant="body2"
            >
              {data.initialFormatted}
            </Typography>
            <Typography variant="body1">{data.finalFormatted}</Typography>
          </div>
        </div>
      );
    } else {
      return (
        <div className={this.classnames.element("priceDetailsContainer")}>
          <Typography
            className={this.classnames.element("finalPrice")}
            variant="body1"
          >
            {data.finalFormatted}
          </Typography>
        </div>
      );
    }
  }

  render() {
    return (
      <div className={this.classnames.root()}>
        <div className={this.classnames.element("priceContainer")}>
          {this.getPrice()}
        </div>
      </div>
    );
  }
}
PriceWidget.classnames = createClassNameHelper(
  `${JSS_CLASS_NAME_PREFIX}PriceWidget`
);
PriceWidget.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string,
    priceContainer: PropTypes.string,
    priceContainerWithSmScreen: PropTypes.string,
    priceDetailsContainer: PropTypes.string,
    priceDetailsContainerWithDiscount: PropTypes.string,
    discountPercent: PropTypes.string,
    discountPercentSkinny: PropTypes.string,
    initialPrice: PropTypes.string,
    finalPrice: PropTypes.string,
    priceDetails: PropTypes.string,
    priceFree: PropTypes.string
  }),
  // price data
  data: PropTypes.shape({
    freeToPlay: PropTypes.bool,
    onSale: PropTypes.bool,
    discountPercentage: PropTypes.number,
    initialFormatted: PropTypes.string,
    finalFormatted: PropTypes.string
  }).isRequired,
  // determines if the widget has 'skinny' styling or not
  skinny: PropTypes.bool
};
PriceWidget.defaultProps = {
  classes: {},
  skinny: false
};

export default withStyles(styles)(PriceWidget);
