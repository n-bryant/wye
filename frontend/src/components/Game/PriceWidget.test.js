import { PriceWidget } from "./PriceWidget";

describe("PriceWidget", () => {
  const props = {
    classes: {
      root: "root",
      priceContainer: "priceContainer",
      priceContainerWithSmScreen: "priceContainerWithSmScreen",
      priceDetailsContainer: "priceDetailsContainer",
      priceDetailsContainerWithDiscount: "priceDetailsContainerWithDiscount",
      discountPercent: "discountPercent",
      discountPercentSkinny: "discountPercentSkinny",
      initialPrice: "initialPrice",
      finalPrice: "finalPrice",
      priceDetails: "priceDetails",
      priceFree: "priceFree"
    },
    data: {
      freeToPlay: true,
      onSale: true,
      discountPercentage: 50,
      initialFormatted: "$10.00",
      finalFormatted: "$5.00"
    },
    skinny: false
  };

  it("should render successfully", () => {
    const wrapper = shallow(<PriceWidget {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("getPrice", () => {
    it("should render a price details container with sale details if the game is on sale", () => {
      const propsWithDiscount = {
        ...props,
        data: {
          ...props.data,
          freeToPlay: false
        }
      };
      const wrapperWithDiscount = shallow(
        <PriceWidget {...propsWithDiscount} />
      );
      const priceDetailsContainer = wrapperWithDiscount.findWhere(n =>
        n.hasClass(props.classes.priceDetailsContainerWithDiscount)
      );
      expect(priceDetailsContainer.length).toBe(1);
      expect(toJson(priceDetailsContainer)).toMatchSnapshot();
    });

    it("should render a price details container with the sole final price value if the game is not on sale", () => {
      const propsWithoutDiscount = {
        ...props,
        data: {
          ...props.data,
          freeToPlay: false,
          onSale: false
        }
      };
      const wrapperWitouthDiscount = shallow(
        <PriceWidget {...propsWithoutDiscount} />
      );
      const priceDetailsContainer = wrapperWitouthDiscount.findWhere(n =>
        n.hasClass(props.classes.priceDetailsContainer)
      );
      expect(priceDetailsContainer.length).toBe(1);
      expect(toJson(priceDetailsContainer)).toMatchSnapshot();
    });

    it("should render a skinny version of the price widget if the skinny prop is truthy", () => {
      const propsWithSkinny = {
        ...props,
        data: {
          ...props.data,
          freeToPlay: false
        },
        skinny: true
      };
      const wrapperSkinny = shallow(<PriceWidget {...propsWithSkinny} />);
      const discountPercentSkinny = wrapperSkinny.findWhere(n =>
        n.hasClass(props.classes.discountPercentSkinny)
      );
      expect(discountPercentSkinny.length).toBe(1);
    });
  });
});
