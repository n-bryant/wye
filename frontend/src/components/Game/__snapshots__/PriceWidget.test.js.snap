// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PriceWidget getPrice should render a price details container with sale details if the game is on sale 1`] = `
<div
  className="wye-PriceWidget__priceDetailsContainer wye-PriceWidget__priceDetailsContainer--withDiscount priceDetailsContainer priceDetailsContainerWithDiscount"
>
  <WithStyles(ForwardRef(Typography))
    className="wye-PriceWidget__discountPercent discountPercent"
    variant="h1"
  >
    -
    50
    %
  </WithStyles(ForwardRef(Typography))>
  <div
    className="wye-PriceWidget__priceDetails priceDetails"
  >
    <WithStyles(ForwardRef(Typography))
      className="wye-PriceWidget__initialPrice initialPrice"
      variant="body2"
    >
      $10.00
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      className="wye-PriceWidget__finalWithSale finalWithSale"
      variant="body1"
    >
      $5.00
    </WithStyles(ForwardRef(Typography))>
  </div>
</div>
`;

exports[`PriceWidget getPrice should render a price details container with the sole final price value if the game is not on sale 1`] = `
<div
  className="wye-PriceWidget__priceDetailsContainer priceDetailsContainer"
>
  <WithStyles(ForwardRef(Typography))
    className="wye-PriceWidget__finalPrice finalPrice"
    variant="body1"
  >
    $5.00
  </WithStyles(ForwardRef(Typography))>
</div>
`;

exports[`PriceWidget should render successfully 1`] = `
<div
  className="wye-PriceWidget root"
>
  <div
    className="wye-PriceWidget__priceContainer priceContainer"
  >
    <WithStyles(ForwardRef(Typography))
      className="wye-PriceWidget__priceFree priceFree"
      variant="h2"
    >
      Free to Play
    </WithStyles(ForwardRef(Typography))>
  </div>
</div>
`;
